scanpipe/pipes/d2d.py

```diff
@@ -0,0 +1,389 @@
+    if logger:
+        logger(f"Mapping {resource_count:,d} .class resources to .java")
+
+    resource_iterator = to_resources_dot_class.iterator(chunk_size=2000)
+    last_percent = 0
+    start_time = timer()
+    for resource_index, to_resource in enumerate(resource_iterator):
+        last_percent = pipes.log_progress(
+            logger,
+            resource_index,
+            resource_count,
+            last_percent,
+            increment_percent=10,
+            start_time=start_time,
+        )
+        _resource_java_to_class_map(to_resource, from_resources)
+
+    # Flag not mapped .class in to/ codebase
+    to_resources_dot_class = to_resources.filter(name__endswith=".class")
+    to_resources_dot_class.update(status=flag.NO_JAVA_SOURCE)
+
+
+def _resource_jar_to_source_map(jar_resource, to_resources, from_resources):
+    jar_extracted_path = get_extracted_path(jar_resource)
+    jar_extracted_files = to_resources.filter(path__startswith=jar_extracted_path)
+
+    # Flag all META-INF/* file as ignored
+    meta_inf_files = jar_extracted_files.filter(path__contains="META-INF/")
+    meta_inf_files.no_status().update(status=flag.IGNORED_META_INF)
+
+    dot_class_files = jar_extracted_files.filter(name__endswith=".class")
+    # Do not continue if some .class files couldn't be mapped.
+    if dot_class_files.has_no_relation().exists():
+        return
+
+    java_to_class_relations = CodebaseRelation.objects.filter(
+        to_resource__in=dot_class_files, map_type="java_to_class"
+    )
+    from_source_roots = [
+        relation.extra_data.get("from_source_root", "")
+        for relation in java_to_class_relations
+    ]
+    if len(set(from_source_roots)) != 1:
+        # Could not determine a common root directory for the java_to_class files
+        return
+
+    try:
+        common_from_resource = from_resources.get(path=from_source_roots[0].rstrip("/"))
+    except ObjectDoesNotExist:
+        return
+
+    pipes.make_relation(
+        from_resource=common_from_resource,
+        to_resource=jar_resource,
+        map_type="jar_to_source",
+    )
+
+
+def jar_to_source_map(project, logger=None):
+    project_files = project.codebaseresources.files()
+    # Include the directories to map on the common source
+    from_resources = project.codebaseresources.from_codebase()
+    to_resources = project_files.to_codebase()
+    to_jars = to_resources.filter(extension=".jar")
+
+    to_jars_count = to_jars.count()
+    if logger:
+        logger(
+            f"Mapping {to_jars_count:,d} .jar resources using jar_to_source_map "
+            f"against from/ codebase"
+        )
+
+    for jar_resource in to_jars:
+        _resource_jar_to_source_map(jar_resource, to_resources, from_resources)
+
+
+def get_diff_ratio(to_resource, from_resource):
+    if not (to_resource.is_text and from_resource.is_text):
+        return
+
+    try:
+        to_lines = to_resource.location_path.read_text().split("\n")
```

Comment:
Consider this, in case we have Windows lines:
```suggestion
        to_lines = to_resource.location_path.read_text().splitlines(True)
```